src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -805,6 +805,8 @@ AdaptiveExecutor(CitusScanState *scanState)
 	TupleDestination *defaultTupleDest =
 		CreateTupleStoreTupleDest(scanState->tuplestorestate, tupleDescriptor);
 
+	bool localExecutionSupported = true;
```

Comment:
It is because of how the code is organized.
 
There are two code paths that can lead to AdaptiveExecutor function:
1.` EXPLAIN ANALYZE select * from ref_table;`
    Goes through the utility hook 
    In **ProcessUtilityInternal** 
        we set EnableLocalExecution = false

> https://github.com/citusdata/citus/blob/efd41e8ea55cf613fc1a0255a034ce13ceb554d3/src/backend/distributed/commands/utility_hook.c#L428-L446:~:text=DisableLocalExecution()%3B

       
2.` 'LOAD 'auto_explain'; `
    `select * from ref_table; `

    Does not go through the utility hook. Explain analyze is requested by AutoExplain implicitly by setting flags in QueryDesc.

    EnableLocalExecution = true // stays with default value


In  AdaptiveExecutor, another flag is used `localExecutionSupported `to disable local execution along with `EnableLocalExecution`.  

https://github.com/citusdata/citus/blob/343d1c5072e652902a91c58ef9538df9a5aa9048/src/backend/distributed/executor/adaptive_executor.c#:~:text=if%20(execution%2D%3E,ShouldExecuteTasksLocally(taskList))

The issue is that `localExecutionSupported `is not set properly probably due to an oversight. Otherwise, the code works as expected.

Reply:
In that case, do we still need `DisableLocalExecution();` call in the code that I shared?

This seems like a more central place for disabling local execution?

Also, what is the difference between `DisableLocalExecution();`  vs `bool localExecutionSupported = false `. Is there a reason that the other code uses the former, and this one uses the latter?

Reply:
BTW, alternatively we can use 
   `DisableLocalExecution();  `
instead of
`  localExecutionSupported = false;`

Reply:
Looks like DisableLocalExecution();  in the utility hook could be removed. 

DisableLocalExecution() is manipulating the GUC value that is usually set explicitly by the client. 
Whereas localExecutionSupported flag is entirely internal.

Reply:
CI run shows that 

 EXPLAIN (*) queries run locally when we remove the DisableLocalExecution() in utility hook. 

 We only catch EXPLAIN ANALYZE in adaptive executor.

Reply:
@onderkalaci removing DisableLocalExecution(); in utility hook did not work as I expected. Pls hold on until I update the PR.

Reply:
> In that case, do we still need `DisableLocalExecution();` call in the code that I shared?
> 
> This seems like a more central place for disabling local execution?
> 
> Also, what is the difference between `DisableLocalExecution();` vs `bool localExecutionSupported = false `. Is there a reason that the other code uses the former, and this one uses the latter?


|     | Utility Hook  | Adaptive Executor |
|--------|--------|--------|
| EXPLAIN stmt | X |  |
| EXPLAIN ANALYZE  stmt   | X | X |
| stmt (/auto_explain on) |  | X | 

When we remove DisableLocalExecution(); in utility hook, EXPLAIN statements will be executed locally. Since the "stmt" is not executed but only explained, I believe this will not cause a problem.

However, the catch is that this is a change in behaviour. For example, in a transaction EXPLAIN will not turn off local execution anymore.

This is the cause of CI failures for the following type of tests.

```BEGIN;
EXPLAIN (COSTS OFF) INSERT INTO numbers SELECT a FROM squares WHERE a < 3;
INSERT INTO numbers SELECT a FROM squares WHERE a < 3;
SELECT * FROM numbers ORDER BY a;
ROLLBACK;
```

I am inclined to keep DisableLocalExecution() as before. What do you think?